File: airdata/validation.js
const { runBash } = require("../shared/helpers")
const getAirData = require("./getAirData")

async function validate() {
  const bbox = [52.46165771962076, 13.278694152832031, 52.5417022642097, 13.485031127929688]
  const dates = ["01.02.2020", "13.02.2020", "19.02.2020", "03.03.2020"]
  const directory = "validation2"

  const allFiles = []
  for (const date of dates) {
    const options = {
      pollutant: "PM10",
      bbox,
      date,
      timestep: 60,
      output: directory,
    }
    const files = await getAirData(options)
    allFiles.push(...files)
  }

  await runBash(`python validation.py --directory=${directory}`)
}

async function test() {
  const options = {
    pollutant: "PM10",
    bbox: [52.25639, 12.874603, 52.778678, 13.932037],
    date: "02.02.2020",
    timestep: 60,
    output: "validation-02-02-2020",
  }
  const filenames = await getAirData(options)

  console.log(filenames)

  const methods = [
    // "nearest-neighbor",
    // "natural-neighbor",
    "idw",
    // "linear-rbf",
    // "thin-plate-rbf",
    // "kriging",
  ]

  const zones = [0, 20, 35, 50, 100, 1000]

  for (const method of methods) {
    for (const file of filenames) {
      const dirName = `${options.output}/${method}`
      await runBash([
        "python interpolate.py",
        `--measurements=${file}`,
        `--method=${method}`,
        `--output=${dirName}`,
        `--zones=${zones.join(",")}`,
        "--visualize=true",
      ])
    }
  }
}

async function download() {
  const dates = [
    // 2nd week of June 2019
    "10.06.2019",
    "11.06.2019",
    "12.06.2019",
    "13.06.2019",
    "14.06.2019",
    "15.06.2019",
    "16.06.2019",
    // 2nd week of August 2019
    "05.08.2019",
    "06.08.2019",
    "07.08.2019",
    "08.08.2019",
    "09.08.2019",
    "10.08.2019",
    "11.08.2019",
    // 2nd week of February 2020
    "10.02.2020",
    "11.02.2020",
    "12.02.2020",
    "13.02.2020",
    "14.02.2020",
    "15.02.2020",
    "16.02.2020",
    // 2nd week of March 2020
    "09.03.2020",
    "10.03.2020",
    "11.03.2020",
    "12.03.2020",
    "13.03.2020",
    "14.03.2020",
    "15.03.2020",
    // 2nd week of April 2020
    "06.04.2020",
    "07.04.2020",
    "08.04.2020",
    "09.04.2020",
    "10.04.2020",
    "11.04.2020",
    "12.04.2020",
  ]

  for (const date of dates) {
    const options = {
      pollutant: "PM10",
      bbox: [52.25639, 12.874603, 52.778678, 13.932037],
      date,
      timestep: 60,
      output: date,
    }
    const files = await getAirData(options)
  }
}

test()
// validate()
// download()

File: airdata/validation.py
import random, pprint, glob
from functools import partial
from argparse import ArgumentParser
import geopandas as gpd
import numpy as np
from scipy.spatial import cKDTree
from sklearn.model_selection import KFold
from sklearn.metrics import mean_squared_error
import matplotlib.pyplot as plt

import interpolators

plt.rcParams["figure.figsize"] = 30, 20
plt.rcParams["font.size"] = 20
plt.rcParams["axes.titlesize"] = 50
plt.rcParams["axes.titlepad"] = 80

interpolator_functions = {
    "nearest-neighbor": interpolators.nearest_neighbor,
    "natural-neighbor": interpolators.discrete_natural_neighbor,
    "idw": interpolators.inverse_distance_weighting,
    "linear-rbf": partial(interpolators.radial_basis_function, function="linear"),
    "mq-rbf": partial(interpolators.radial_basis_function, function="multiquadric"),
    "imq-rbf": partial(
        interpolators.radial_basis_function, function="inverse-multiquadric"
    ),
    "thin-plate-rbf": partial(
        interpolators.radial_basis_function, function="thin-plate"
    ),
    # "kriging": interpolators.kriging,
}


def cross_validate_file(measurements_file):
    measurements = gpd.read_file(measurements_file)

    external_crs = "EPSG:4326"
    internal_crs = "EPSG:3068"

    measurements = measurements.to_crs(internal_crs)

    x = np.array(measurements.geometry.x)
    y = np.array(measurements.geometry.y)
    values = np.array(measurements.value)

    xmin, ymin, xmax, ymax = measurements.total_bounds
    size = 100  # grid cell size in meters
    xnew = np.linspace(xmin, xmax, int((xmax - xmin) / size))
    ynew = np.linspace(ymin, ymax, int((ymax - ymin) / size))

    folds = 10
    seed = random.randint(0, 9999)
    kfold = KFold(folds, True, seed)
    avg_rmse_per_method = {}
    for method in interpolator_functions:
        interpolation_method = interpolator_functions[method]
        sum_rmse = 0
        for train, test in kfold.split(values):
            train_points = np.column_stack((x[train], y[train]))
            train_values = values[train]

            interpolated_values = interpolation_method(
                x[test], y[test], train_points, train_values, grid=False
            )
            rmse = mean_squared_error(values[test], interpolated_values)
            sum_rmse += rmse

        avg_rmse = sum_rmse / folds
        avg_rmse_per_method[method] = avg_rmse

    return avg_rmse_per_method


# cross_validate_file("./validation-02-02-2020/data_2020-02-02T08-00-00.geojson")


def cross_validation(directory):
    files = glob.glob(f"{directory}/*.geojson")

    win_counter = dict.fromkeys(interpolator_functions.keys(), 0)
    for file in files:
        result = cross_validate_file(file)
        # pprint.pprint(result)
        winning_method = min(result, key=result.get)
        win_counter[winning_method] += 1

    pprint.pprint(win_counter)
    fig, ax = plt.subplots()
    x_pos = np.arange(0, len(win_counter.keys()))
    heights = list(win_counter.values())
    labels = list(win_counter.keys())

    ax.bar(x_pos, heights)
    ax.set_xticks(x_pos)
    ax.set_xticklabels(labels)
    ax.set_ylabel("No. of wins")
    ax.set_title("Interpolation method cross-validation")

    for i, h in enumerate(heights):
        margin = np.max(heights) * 0.02
        ax.text(
            x=i, y=h + margin, s=f"{h}", va="center", color="black", fontweight="bold"
        )

    fig.savefig(f"{directory}/wins.png")


parser = ArgumentParser()
parser.add_argument(
    "--directory",
    dest="directory",
    help="Directory to all GeoJSON measurement files",
    metavar="FILE",
)
args = parser.parse_args()

cross_validation(args.directory)
